test/extensions/clusters/common/logical_host_integration_test.cc

```diff
@@ -0,0 +1,99 @@
+                                   public HttpIntegrationTest {
+public:
+  LogicalHostIntegrationTest()
+      : HttpIntegrationTest(Http::CodecType::HTTP1, GetParam()),
+        registered_dns_factory_(dns_resolver_factory_) {}
+
+  void createUpstreams() override { HttpIntegrationTest::createUpstreams(); }
+
+  NiceMock<Network::MockDnsResolverFactory> dns_resolver_factory_;
+  Registry::InjectFactory<Network::DnsResolverFactory> registered_dns_factory_;
+  uint32_t address_ = 0;
+  std::string address_string = "::1";
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, LogicalHostIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// Reproduces a race from https://github.com/envoyproxy/envoy/issues/32850.
+// The test is by mocking the DNS resolver to return multiple different
+// addresses, also config dns_refresh_rate to be extremely fast.
+TEST_P(LogicalHostIntegrationTest, DISABLED_FastChangingDnsResult) {
+  // Only needs to test the race with one stack.
+  if (version_ == Network::Address::IpVersion::v4) {
+    return;
+  }
+
+  auto dns_resolver = std::make_shared<Network::MockDnsResolver>();
+  EXPECT_CALL(dns_resolver_factory_, createDnsResolver(_, _, _))
+      .WillRepeatedly(testing::Return(dns_resolver));
+  EXPECT_CALL(*dns_resolver, resolve(_, _, _))
+      .WillRepeatedly(
+          Invoke([&](const std::string&, Network::DnsLookupFamily,
+                     Network::DnsResolver::ResolveCb dns_callback) -> Network::ActiveDnsQuery* {
+            // Return multiple mixed v4/v6 addresses to increase the race window.
+            dns_callback(Network::DnsResolver::ResolutionStatus::Success,
+                         TestUtility::makeDnsResponse({
+                             "::1",
+                             absl::StrCat("127.0.0.", address_),
+                             absl::StrCat("127.0.0.", address_ + 1),
+                             absl::StrCat("127.0.0.", address_ + 2),
+                             absl::StrCat("127.0.0.", address_ + 3),
+                             absl::StrCat("127.0.0.", address_ + 4),
+                             absl::StrCat("127.0.0.", address_ + 5),
+                             absl::StrCat("127.0.0.", address_ + 6),
+                             absl::StrCat("127.0.0.", address_ + 7),
+                             absl::StrCat("127.0.0.", address_ + 8),
+                             absl::StrCat("127.0.0.", address_ + 9),
+                             "::2",
+                             "::3",
+                             "::4",
+                             "::5",
+                             "::6",
+                             "::7",
+                             "::8",
+                             "::9",
+                         }));
+            address_ = (address_ + 1) % 128;
```

Comment:
Actually the addresses after first_address_string_ does not mean that much. They won't be used in creating connections.  The reason we want to change them during consecutive DNS refresh is because this check will be true:
https://github.com/envoyproxy/envoy/blob/002e7f9cd0224800e5b318cd7d84c49fb65b9565/source/extensions/clusters/logical_dns/logical_dns_cluster.cc#L136

Thus

logical_host_->setNewAddress() will be called:

https://github.com/envoyproxy/envoy/blob/002e7f9cd0224800e5b318cd7d84c49fb65b9565/source/extensions/clusters/logical_dns/logical_dns_cluster.cc#L142

Reply:
can you add a comment to that effect?

Reply:
I have below comments: 
// Keep changing the returned addresses to force address update.

You mean to explicitly mention the code in the comments? That maybe too verbose?

Reply:
BTW, please suggest the comments you would like to add.

Reply:
Suggest:
```
// The only signficant address is the first one; the other ones are just used to populate a list
// whose maintenance is race-prone.
```

Reply:
done